File: linechart/src/main/java/com/davidmiguel/linechart/LineChart.java
package com.davidmiguel.linechart;

import android.graphics.Path;

import com.davidmiguel.linechart.animation.LineChartAnimator;
import com.davidmiguel.linechart.touch.OnScrubListener;

import java.util.List;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

@SuppressWarnings("unused")
public interface LineChart {

    /**
     * Get the backing {@link LineChartAdapter}.
     */
    @Nullable
    LineChartAdapter getAdapter();

    /**
     * Sets the backing {@link LineChartAdapter} to generate the points to be graphed.
     */
    void setAdapter(@Nullable LineChartAdapter adapter);

    /**
     * Gets the color of the line.
     */
    @ColorInt
    int getLineColor();

    /**
     * Sets the color of the line.
     */
    void setLineColor(@ColorInt int lineColor);

    /**
     * Gets the width in pixels of the line's stroke.
     */
    float getLineWidth();

    /**
     * Sets the width in pixels of the line's stroke.
     */
    void setLineWidth(float lineWidth);

    /**
     * Gets the corner radius in pixels used when rounding the line's segments.
     */
    float getCornerRadius();

    /**
     * Sets the corner radius in pixels to use when rounding the line's segments.
     * Passing 0 indicates that corners should not be rounded.
     */
    void setCornerRadius(float cornerRadius);

    /**
     * Gets how the chart has to be filled if fill is enabled.
     */
    @LineChartFillType
    int getFillType();

    /**
     * Sets how the chart has to be filled if fill is enabled.
     */
    void setFillType(@LineChartFillType int fillType);

    /**
     * Gets the color of the filling.
     */
    @ColorInt
    int getFillColor();

    /**
     * Sets the color of the filling.
     */
    void setFillColor(@ColorInt int fillColor);

    /**
     * Gets the color of the grid line.
     */
    @ColorInt
    int getGridLineColor();

    /**
     * Sets the color of the base line.
     */
    void setGridLineColor(@ColorInt int gridLineColor);

    /**
     * Gets the width in pixels of the grid line's stroke.
     */
    float getGridLineWidth();

    /**
     * Sets the width in pixels of the grid line's stroke.
     */
    void setGridLineWidth(float gridLineWidth);

    /**
     * Gets the number of grid divisions in X axis.
     */
    int getGridXDivisions();

    /**
     * Sets the number of grid divisions in X axis.
     */
    void setGridXDivisions(int gridXDivisions);

    /**
     * Gets the number of grid divisions in Y axis.
     */
    int getGridYDivisions();

    /**
     * Sets the number of grid divisions in Y axis.
     */
    void setGridYDivisions(int gridYDivisions);

    /**
     * Gets the color of the base line.
     */
    @ColorInt
    int getBaseLineColor();

    /**
     * Sets the color of the base line.
     */
    void setBaseLineColor(@ColorInt int baseLineColor);

    /**
     * Gets the width in pixels of the base line's stroke.
     */
    float getBaseLineWidth();

    /**
     * Sets the width in pixels of the base line's stroke.
     */
    void setBaseLineWidth(float baseLineWidth);

    /**
     * Returns true if scrubbing is enabled on this view.
     */
    boolean isScrubEnabled();

    /**
     * Sets whether or not to enable scrubbing on this view.
     */
    void setScrubEnabled(boolean scrubbingEnabled);

    /**
     * Gets the color of the scrub line.
     */
    @ColorInt
    int getScrubLineColor();

    /**
     * Sets the color of the scrub line.
     */
    void setScrubLineColor(@ColorInt int scrubLineColor);

    /**
     * Gets the width in pixels of the scrub line's stroke.
     */
    float getScrubLineWidth();

    /**
     * Set the width in pixels of the scrub line's stroke.
     */
    void setScrubLineWidth(float scrubLineWidth);

    /**
     * Gets the current {@link OnScrubListener}.
     */
    @Nullable
    OnScrubListener getScrubListener();

    /**
     * Sets a {@link OnScrubListener} to be notified of the user's scrubbing gestures.
     */
    void setScrubListener(@Nullable OnScrubListener scrubListener);

    /**
     * Gets the Animator class used to animate the Line Chart.
     */
    @Nullable
    LineChartAnimator getLineChartAnimator();

    /**
     * Animator class to animate the Line Chart.
     */
    void setLineChartAnimator(@Nullable LineChartAnimator lineChartAnimator);

    /**
     * Sets the path to animate in onDraw, used by animators.
     */
    void setAnimationPath(@NonNull Path animationPath);

    /**
     * Returns a copy of current graphic X points.
     */
    @NonNull
    List<Float> getXPoints();

    /**
     * Returns a copy of current graphic Y points.
     */
    @NonNull
    List<Float> getYPoints();
}

File: linechart/src/main/java/com/davidmiguel/linechart/LineChart.kt
package com.davidmiguel.linechart

import android.graphics.Path
import android.graphics.RectF
import androidx.annotation.ColorInt
import com.davidmiguel.linechart.animation.LineChartAnimator
import com.davidmiguel.linechart.formatter.YAxisValueFormatter
import com.davidmiguel.linechart.touch.OnScrubListener

interface LineChart {

    // Data

    /**
     * Sets the backing [LineChartAdapter] to generate the points to be graphed.
     */
    var adapter: LineChartAdapter

    /**
     * Rectangle where the chart is painted.
     */
    val drawingArea: RectF

    /**
     * Returns a copy of current graphic X points.
     */
    val xPoints: List<Float>

    /**
     * Returns a copy of current graphic Y points.
     */
    val yPoints: List<Float>

    // Styleable properties

    /**
     * Sets the color of the line.
     */
    @get:ColorInt
    var lineColor: Int

    /**
     * Sets the width in pixels of the line's stroke.
     */
    var lineWidth: Float

    /**
     * Sets the corner radius in pixels to use when rounding the line's segments.
     * Passing 0 indicates that corners should not be rounded.
     */
    var cornerRadius: Float

    /**
     * Sets how the chart has to be filled if fill is enabled.
     */
    @get:LineChartFillType
    var fillType: Int

    /**
     * Sets the color of the filling.
     */
    @get:ColorInt
    var fillColor: Int

    /**
     * Sets the color of the base line.
     */
    @get:ColorInt
    var gridLineColor: Int

    /**
     * Sets the width in pixels of the grid line's stroke.
     */
    var gridLineWidth: Float

    /**
     * Sets the color of the base line.
     */
    @get:ColorInt
    var baseLineColor: Int

    /**
     * Sets the width in pixels of the base line's stroke.
     */
    var baseLineWidth: Float

    /**
     * Sets the color of the zero line.
     */
    @get:ColorInt
    var zeroLineColor: Int

    /**
     * Sets the width in pixels of the zero line's stroke.
     */
    var zeroLineWidth: Float

    /**
     * Sets the side margin of the labels.
     */
    var labelMargin: Float

    /**
     * Sets the text color of the labels.
     */
    @get:ColorInt
    var labelTextColor: Int

    /**
     * Sets the text size of the labels.
     */
    var labelTextSize: Float

    /**
     * Sets the background color of the labels.
     */
    @get:ColorInt
    var labelBackgroundColor: Int

    /**
     * Sets the background radius of the labels.
     */
    var labelBackgroundRadius: Float

    /**
     * Sets the background horizontal padding of the labels.
     */
    var labelBackgroundPaddingHorizontal: Float

    /**
     * Sets the background vertical padding of the labels.
     */
    var labelBackgroundPaddingVertical: Float

    /**
     * Sets the text color of the zero label.
     */
    @get:ColorInt
    var zeroLabelTextColor: Int

    /**
     * Sets the text size of the zero label.
     */
    var zeroLabelTextSize: Float

    /**
     * Sets the background color of the zero label.
     */
    @get:ColorInt
    var zeroLabelBackgroundColor: Int

    /**
     * Sets the color of the scrub line.
     */
    @get:ColorInt
    var scrubLineColor: Int

    /**
     * Set the width in pixels of the scrub line's stroke.
     */
    var scrubLineWidth: Float

    // How to draw

    /**
     * Sets formatter used to format labels in Y axis.
     * If null is passed, the default formatter will be used.
     */
    var yAxisValueFormatter: YAxisValueFormatter

    // What to draw

    /**
     * Sets the number of grid divisions in X axis.
     */
    var gridXDivisions: Int

    /**
     * Sets the number of grid divisions in Y axis.
     */
    var gridYDivisions: Int

    /**
     * Sets whether or not to show the zero line.
     */
    var isZeroLineEnabled: Boolean

    /**
     * Sets whether or not to enable scrubbing on this view.
     */
    var isScrubEnabled: Boolean

    // Touch

    /**
     * Sets a [OnScrubListener] to be notified of the user's scrubbing gestures.
     */
    var scrubListener: OnScrubListener?

    // Animation

    /**
     * Animator class to animate the Line Chart.
     */
    var lineChartAnimator: LineChartAnimator?

    /**
     * Sets the path to animate in onDraw, used by animators.
     */
    var animationPath: Path?
}
